pynicotine/plugins/leech_detector/__init__.py

```diff
@@ -71,72 +76,90 @@ def loaded_notification(self):
         if self.settings["num_folders"] < min_num_folders:
             self.settings["num_folders"] = min_num_folders
 
-        if self.settings["message"]:
-            self.str_action = "message leecher"
-        else:
-            self.str_action = "log leecher"
+        # Separate leechers set for faster membership checks
+        self.detected_leechers = set(self.settings["detected_leechers"])
 
         self.log(
-            "Ready to %ss, require users have a minimum of %d files in %d shared public folders.",
-            (self.str_action, self.settings["num_files"], self.settings["num_folders"])
+            "Require users have a minimum of %d files in %d shared public folders.",
+            (self.settings["num_files"], self.settings["num_folders"])
         )
 
     def upload_queued_notification(self, user, virtual_path, real_path):
 
-        if user in self.probed:
+        if user in self.probed_users:
             # We already have stats for this user.
             return
 
-        self.probed[user] = "requesting"
+        self.probed_users[user] = "requesting"
         self.core.request_user_stats(user)
+
         self.log("Getting statistics from the server for new user %s…", user)
 
     def user_stats_notification(self, user, stats):
 
-        if user not in self.probed:
+        if user not in self.probed_users:
             # We did not trigger this notification
             return
 
-        if self.probed[user] != "requesting":
+        if self.probed_users[user] != "requesting":
             # We already dealt with this user.
             return
 
-        if stats["files"] >= self.settings["num_files"] and stats["dirs"] >= self.settings["num_folders"]:
-            self.log("User %s is okay, sharing %s files in %s folders.", (user, stats["files"], stats["dirs"]))
-            self.probed[user] = "okay"
+        num_files = stats["files"]
+        num_folders = stats["dirs"]
+        is_previous_leecher = (user in self.detected_leechers)
+
+        if num_files >= self.settings["num_files"] and num_folders >= self.settings["num_folders"]:
+            if is_previous_leecher:
```

Comment:
This code is unreachable, it is not above the condition on Line 104.

Reply:
It's reachable in later sessions. Not sure if it's worth making it reachable in the same session, it would require adding some checks to ensure we don't spam the server with user stats requests.

Reply:
It is still unreachable, due to the condition on Line 89, consider in `upload_queued_notification()`:

```
    if user in self.probed_users and user not in self.detected_leechers:
        # We already have stats and accepted this user
```

Reply:
I'll revisit later, this fucking plugin is getting on my nerves.

Reply:
I'm not going to spam the server with user stat requests on every queued file request, we need something better.

Reply:
Alright, pushed a change that should make my idea work.